# Part 1. Equation of a Slime

How many late days are you using for this assignment?

3

In [ ]:
# Imports section
%pip install pandas
%pip install scikit-learn
%pip install numpy

import pandas as pd
import sklearn as sk
import numpy as np

## 1. Loading the dataset

In [2]:
# Using pandas load the dataset

data = pd.read_csv('science_data_large.csv')

# Output the first 15 rows of the data

print(data.head(15))

# Display a summary of the table information (number of datapoints, etc.)

data.info()

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)

X = data[['Temperature °C', 'Mols KCL']]
y = data['Size nm^3']

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1 , random_state=42)

## 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set

model = sk.linear_model.LinearRegression()
model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model

sample_datapoint = pd.DataFrame([[413, 612]], columns=['Temperature °C', 'Mols KCL'])
predicted_output = model.predict(sample_datapoint)
print(predicted_output)

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

model_score = model.score(X_test, y_test)
print(model_score)

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

coefficients = model.coef_
intercept = model.intercept_

print(coefficients)
print(intercept)

[580336.3698285]
0.8552472077276096
[ 866.14641337 1032.69506649]
-409391.4795834081


Write the linear equation of a slime: (example equation: $E = mc^2$)

$h(x) = -409391.47958 + 866.14641 * x_1 + 1032.69506 * x_2$

Report on score and explain meaning:

By default, the score method for LinearRegression calculates the R^2 value to score the model. The R^2 value divides the residual sum of squares (the variance of the target variable not accounted for by the model) b the total sum of squares (the total variance of the target variable) and subtracts it from 1. This means that the closer the R^2 value is to 1, the less the fraction is, which translates to less variance of the target variable not account for by the model and thus a more accurate prediction. The model above has an R^2 value of 0.85524, which is overall pretty good. 

## 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores = sk.model_selection.cross_val_score(model, X, y, cv=kf)  

print(scores)

# Report on their finding and their significance

[0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]


Write findings here:

The R^2 value across shuffles of data remains between 0.82 to 0.89, so within +/- 0.04 range of the score calculated in 1.5. This range is relatively narrow, which suggests that the high score in 1.3 was not a lucky outlier but rather than the model is stable and consistently performs well in predicting the target value. 

## 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

poly = sk.preprocessing.PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

model_poly = sk.linear_model.LinearRegression()
model_poly.fit(X_poly,y)

kf_poly = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_poly = sk.model_selection.cross_val_score(model_poly, X_poly, y, cv=kf_poly)  

print(scores_poly)

coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

print(coefficients_poly)
print(intercept_poly)

# Report on the metrics and output the resultant equation as you did in Part 3.

[1. 1. 1. 1. 1.]
[ 0.00000000e+00  1.20000000e+01 -1.23110897e-07 -1.05662476e-11
  2.00000000e+00  2.85714287e-02]
1.6573118045926094e-05


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

$h(x) = 0.00001 + 12x_1 - 0.00000x_2 - 0.00000x_1^2 + 2x_1x_2 + 0.02857x_2^2$

Report on the score and interpret:

The R^2 values across multiple shuffles all return 1.0, which means that there was no variance of the target variable not accounted for by the model and that the models predictions are perfect. Since the scores remained consistently perfect across shuffles, this suggests that the equation above perfectly represents how temperature and potassium chloride interact to govern the growth of slime. 

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

## Classification Model Evaluation

In [7]:
# Load the dataset. Then train and evaluate the classification models.

data = pd.read_csv('ckd_feature_subset.csv')

X = data[['age', 'bp', 'wbcc', 'appet_poor', 'appet_good', 'rbcc']]
y = data['Target_ckd']

X_train, X_test, y_train, y_test = sk.model_selection.train_test_split(X, y, test_size=0.1 , random_state=42)

### Logistic Regression

In [8]:
model_logistic = sk.linear_model.LogisticRegression()
model_logistic.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_logistic = sk.model_selection.cross_val_score(model_logistic, X, y, cv=kf, scoring='accuracy')

mean_logistic = np.mean(scores_logistic)
std_logistic = np.std(scores_logistic)

print(mean_logistic)
print(std_logistic)

0.8565591397849461
0.0662686697755876


### Support Vector Machines

In [9]:
model_svc = sk.svm.SVC()
model_svc.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_svc = sk.model_selection.cross_val_score(model_svc, X, y, cv=kf, scoring='accuracy')

mean_svc = np.mean(scores_svc)
std_svc = np.std(scores_svc)

print(mean_svc)
print(std_svc)


0.9281720430107526
0.04760126519964333


### K-Nearest Neighbors

In [10]:
model_neighbors = sk.neighbors.KNeighborsClassifier()
model_neighbors.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_neighbors = sk.model_selection.cross_val_score(model_neighbors, X, y, cv=kf, scoring='accuracy')

mean_neighbors = np.mean(scores_neighbors)
std_neighbors = np.std(scores_neighbors)

print(mean_neighbors)
print(std_neighbors)

0.9279569892473118
0.05243961560473918


### Neural Networks

In [21]:
model_nn = sk.neural_network.MLPClassifier()
model_nn.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_nn = sk.model_selection.cross_val_score(model_nn, X, y, cv=kf, scoring='accuracy')

mean_nn = np.mean(scores_nn)
std_nn = np.std(scores_nn)

print(mean_nn)
print(std_nn)

/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.wa

0.8763440860215054
0.06870286199728586


/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

### Classification Experiments Table

| Model                   | Mean Accuracy         | Standard Deviation     |
|-------------------------|-----------------------|-------------------------|
| Logistic Regression      | 0.8565591397849461    | 0.0662686697755876      |
| Support Vector Machines   | 0.9281720430107526    | 0.04760126519964333     |
| K-Nearest Neighbors      | 0.9279569892473118    | 0.05243961560473918     |
| Neural Networks          | 0.8763440860215054    | 0.06870286199728586    |

From the classification experiments, Logistic Regression performed the worse, followed by the Neural Network model. They both had a mean accuracy below 0.9 as well as the highest standard deviations. The SVC model and the K-Nearest Neighbors model were the better of the 4 models explored. Both had very similar mean accuracies and standard deviations, with the SVC model being slightly more accurate with a lower standard deviation. This is likely because SVC models are more effective then others when working with a dataset with a relatively small amount of samples and a large amount of features, as well as a dataset that has non-linear relationships between features. These characteristics are present in the ckd_feature_subset dataset. 

## Experiments with Neural Networks

### Hidden Layer Sizes

In [22]:
model_nn_layer = sk.neural_network.MLPClassifier(max_iter=1000)
model_nn_layer.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_nn_layer = sk.model_selection.cross_val_score(model_nn_layer, X, y, cv=kf, scoring='accuracy')

mean_nn_layer = np.mean(scores_nn_layer)
std_nn_layer = np.std(scores_nn_layer)

print(mean_nn_layer)
print(std_nn_layer)

0.9350537634408603
0.040465552982107675


### Solver

In [29]:
model_nn_solver = sk.neural_network.MLPClassifier(solver='lbfgs')
model_nn_solver.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_nn_solver = sk.model_selection.cross_val_score(model_nn_solver, X, y, cv=kf, scoring='accuracy')

mean_nn_solver = np.mean(scores_nn_solver)
std_nn_solver = np.std(scores_nn_solver)

print(mean_nn_solver)
print(std_nn_solver)

0.9804301075268818
0.0259173245508522


### Activation

In [32]:
model_nn_activation = sk.neural_network.MLPClassifier(activation='tanh')
model_nn_activation.fit(X_train, y_train)

kf = sk.model_selection.KFold(n_splits=5, shuffle=True, random_state=42)
scores_nn_activation = sk.model_selection.cross_val_score(model_nn_activation, X, y, cv=kf, scoring='accuracy')

mean_nn_activation = np.mean(scores_nn_activation)
std_nn_activation = np.std(scores_nn_activation)

print(mean_nn_activation)
print(std_nn_activation)

/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.wa

0.8825806451612903
0.05966879720722754


/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/Users/<redacted>/Library/Python/3.9/lib/python/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

### Neural Network Experiments Table

| Configurations          | Mean Accuracy         | Standard Deviation     |
|-------------------------|-----------------------|-------------------------|
| Default                 | 0.8763440860215054    | 0.06870286199728586    |
| Hidden Layer Sizes      | 0.9350537634408603    | 0.040465552982107675   |
| Solver                  | 0.9804301075268818    | 0.0259173245508522     |
| Activation              | 0.8825806451612903    | 0.05966879720722754   |

I changed three main configurations: Hidden Layer Sizes, Solver, and Activation. Activation had the least affect, with the mean accuracy and standard deviation being only slightly better than default configurations. Chaing Hidden Layer Sizes to 1000 was the next most effective, providing a mean accuracy and standard deviation that was higher than any of the models in the previous classification experiment. By far the most effective change, however, was solver. By changing the solver to lbfgs, the mean accuracy jumps by more than 0.1 to around 0.98, a near perfect score. It also yielded the least standard deviation.